Spigot-Server-Patches/0680-Remove-streams-from-SensorNearest.patch

```diff
@@ -0,0 +1,117 @@
++        List<EntityLiving> visibleMobs = new java.util.ArrayList<>(list);
++        visibleMobs.removeIf(otherEntityLiving -> !Sensor.a(entityliving, otherEntityLiving));
++        behaviorcontroller.setMemory(MemoryModuleType.VISIBLE_MOBS, visibleMobs);
++        // Paper end
+     }
+ 
+     @Override
+diff --git a/src/main/java/net/minecraft/server/SensorNearestPlayers.java b/src/main/java/net/minecraft/server/SensorNearestPlayers.java
+index 904a6d5ac61d2ac81f1057068383e9ab432852db..5947dc69e52224dfe17eb2e3afc2a26e96a41584 100644
+--- a/src/main/java/net/minecraft/server/SensorNearestPlayers.java
++++ b/src/main/java/net/minecraft/server/SensorNearestPlayers.java
+@@ -19,22 +19,23 @@ public class SensorNearestPlayers extends Sensor<EntityLiving> {
+ 
+     @Override
+     protected void a(WorldServer worldserver, EntityLiving entityliving) {
+-        Stream stream = worldserver.getPlayers().stream().filter(IEntitySelector.g).filter((entityplayer) -> {
+-            return entityliving.a((Entity) entityplayer, 16.0D);
+-        });
++        // Paper start - remove streams in favour of lists
++        List<EntityHuman> players = new java.util.ArrayList<>(worldserver.getPlayers());
++        players.removeIf(player -> IEntitySelector.notSpectator().test(player) || entityliving.a(player, 16.0D)); // Paper - removeIf only re-allocates once compared to iterator
++        players.sort(Comparator.comparingDouble(entityliving::h));
+ 
+-        entityliving.getClass();
+-        List<EntityHuman> list = (List) stream.sorted(Comparator.comparingDouble(entityliving::h)).collect(Collectors.toList());
+         BehaviorController<?> behaviorcontroller = entityliving.getBehaviorController();
+-
+-        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_PLAYERS, (Object) list);
+-        List<EntityHuman> list1 = (List) list.stream().filter((entityhuman) -> {
+-            return a(entityliving, (EntityLiving) entityhuman);
+-        }).collect(Collectors.toList());
+-
+-        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, (Object) (list1.isEmpty() ? null : (EntityHuman) list1.get(0)));
+-        Optional<EntityHuman> optional = list1.stream().filter(IEntitySelector.f).findFirst();
+-
+-        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_TARGETABLE_PLAYER, optional);
++        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_PLAYERS, players);
++
++        EntityHuman nearest = null;
++        for (EntityHuman player : players) {
++            if (!Sensor.a(entityliving, player)) {
++                nearest = player;
++                break;
++            }
++        }
++        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, nearest);
++        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_TARGETABLE_PLAYER, nearest != null && IEntitySelector.canAITarget().test(nearest) ? Optional.of(nearest) : Optional.empty());
```

Comment:
This has a behavioural difference: the stream will find the first that fulfills `IEntitySelector.canAITarget()`; consider instead:

```java
EntityHuman nearest = null, targettable = null;
for (EntityHuman player : players) {
    if (!Sensor.a(entityliving, player)) {
        if (nearest == null) nearest = player;
        if (targettable == null && IEntitySelector.canAITarget().test(player)) targettable = player;
    }
    if (nearest != null && targettable != null) break;
}
```

Reply:
Right, thanks for catching it!